### PR TITLE
more strict checks in analy job broker; compute to-running rate of sites

### PR DESCRIPTION
**Pull request by mightqxc**
The idea is to make analy broker more strict to filter out bad sites. I discussed with Fernando about the idea.

E.g. This task was shown in my slow task report. Slowness was due to jobs stuck on ANALY_DCSC.
Although ANALY_DCSC had zero running jobs and behaved inactive, JEDI still brokered jobs to ANALY_DCSC at that time, since for some input files of the task, ANALY_DCSC was the only site that had the input files (WITH data locality) and also passed all checks:

```
2020-04-08 20:22:05,978 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702> start
2020-04-08 20:22:06,017 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> initial 433 candidates
2020-04-08 20:22:06,025 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702> !!! look for candidates WITH data locality check
2020-04-08 20:22:06,026 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702> getting the list of sites where user.cgottard:user.cgottard.HDBS3.data18_13TeV.periodF_EXT0.304432020 is available
2020-04-08 20:22:06,178 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702>  35 sites : ...
2020-04-08 20:22:06,180 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 18 candidates have input data
2020-04-08 20:22:06,180 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702>   skip site=ANALY_BOINC due to status=test criteria=-status
...
2020-04-08 20:22:06,605 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed memory check = 4350 MB
2020-04-08 20:22:06,606 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> maxAtomSize=219614246 effectiveAtomSize=209.44046592712402 outDiskCount=101376 workDiskSize=2101346304
2020-04-08 20:22:06,606 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> minDiskCountScratch=2233.0 minDiskCountRemote=2024.0 nGBPerJobInMB=None
2020-04-08 20:22:06,606 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed scratch disk check
2020-04-08 20:22:06,606 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed SE space check
2020-04-08 20:22:06,616 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed pilot activity check
2020-04-08 20:22:06,617 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed inclusion/exclusion/cloud
2020-04-08 20:22:06,853 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed hospital check
2020-04-08 20:22:07,071 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> 1 candidates passed overload check
2020-04-08 20:22:07,071 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702> getting the list of available files for user.cgottard:user.cgottard.HDBS3.data18_13TeV.periodF_EXT0.304432020
2020-04-08 20:22:07,410 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> totalInputSize=2 GB
2020-04-08 20:22:07,410 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702> final 1 candidates
2020-04-08 20:22:07,410 panda.log.AtlasAnalJobBroker: INFO     <jediTaskID=20990702>   use site=ANALY_DCSC with weight=0.0006257822277847309 nRunning=0 nDefined=967 nActivated=459 nStarting=159 nAssigned=12 nFailed=0 nClosed=0 nFinished=0 dataW=1 totalInGB=2 localInGB=2 nFiles=11 nLocalDisk=11 nLocalTape=0 nCache=0 nRemote=0 criteria=+use
2020-04-08 20:22:09,451 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702> sent
2020-04-08 20:22:09,451 panda.log.AtlasAnalJobBroker: DEBUG    <jediTaskID=20990702> done
```

This code change will view these kinds of sites as problematic sites and filter them out.

The to-running rate of sites are computed and printed in logs, but not used in algorithm yet. It can be used when adding more filters or adjusting weight in the future.

Still a draft request. Any comment is appreciated.

### CODE REVIEW

**Comment by tmaeno** [comment]
I think that the idea itself is fine. However, the brokerage calls getSiteToRunRateStats() for each combination of task and dataset, which could be expensive for the database. It would be better to use a table periodically updated with aggregated information rather than directly scanning job tables.

**Comment by fbarreir** [comment]
Ideally you should be using atlas_panda.jobs_share_stats that one aggregates the jobsActive4 table every few minutes

**Comment by mightqxc** [comment]
Thanks for your comments. And sure, I will work on performance improvement since the idea is agreed.

> atlas_panda.jobs_share_stats

The to-running rate of site cannot be evaluate with the data in atlas_panda.jobs_share_stats since the table only contains current information, but not historical information (say, how many running jobs 2 hours ago).
Not sure if it's worthy to add another table for this purpose.
Other solution I can think of is to query elasticsearch for the data instead, or cache the query result somewhere to reduce frequency of DB query.

File: pandajedi/jedibrokerage/AtlasBrokerUtils.py [hunk 1]
```diff
@@ -642,6 +643,54 @@ def getOkNgArchList(task_spec):
...
+    # condition of query
+    if cache_key in CACHE_SiteToRunRateStats \
+        and current_time >= CACHE_SiteToRunRateStats[cache_key]['exp']:
+        # query from local cache
+        ret_val = True
+        ret_map = CACHE_SiteToRunRateStats[cache_key]['data']
+    else:
+        if False:
+        # query from DB cache
+        # cached_data = tbIF.getCachedData(key='SiteToRunRateStats')
```
Comment (mightqxc) [inline pandajedi/jedibrokerage/AtlasBrokerUtils.py:671]: If there exists a table in PanDA DB acting like harvester cache_table where arbitrary data can be shared across all nodes, this method can be implemented.

**Comment by mightqxc** [comment]
@tmaeno 
Try to explain my idea with the commit:
Several caches are added to reduce DB load.
If a "cache_table" in PanDA DB exists, I can even add a new watchdog to query job information from elasticsearch and fill into the cache_table for all the nodes. In this way it never needs to query jobs from DB (unless elasticsearch is down) and minimize the need to query over network.

Besides, adding a cache table (for arbitrary content) seems to be more useful and flexible in the future than adding a table to aggregate job information dedicate to this use case.

What do you think? Any comment is appreciated.

**Comment by mightqxc** [comment]
@tmaeno With cache table it reduces much frequency to query job table (once 10 min.). Tested to be OK. Ready to merge.

**Pull request merged**